### PR TITLE
[IMP] crm: redefine won/lost and apply rules in PLS frequencies and views

### PR DESCRIPTION
**Pull request by dbeguin**
BEFORE:

LOST:     0% and not active
WON:    100% and active

NOW:  

LOST:     0% and not active
WON:    stage won and 100% proba (!this should be implied!)

Purpose
=======
Ribbons and buttons display + PLS won/lost frequency rules should be aligned.
 --> Redefine Won / Lost heuristic and apply it everywhere.

Specification
=============

"New heuristic":
Lost: Proba = 0 AND archived (and not in won stage)
Won: won stage AND 100% (was previously: won stage and active)
	(Won stage implies 100%.)

That means that even if a lead is archived, it is still considered as won if
he is in won stage. That also mean that we should block the edition of
probability in won stage as it should always be 100%.
In the same way, we should block edition on probability if the lead is lost.

Won / lost ribbons display rules should be aligned.

That also implies to change a bit the rules to handle won/lost frequencies

- Set to Won Stage (>WS) : Reach Won
	If probability was 0 and active was False: Leave Lost
- Move out from Won stage (<WS) : Leave Won
	If active was False and Proba = 0 -> This should never happen
- Archive: If set proba=0 or proba was 0: Reach Lost
- Restore: Leave Lost

Special cases :
- Out of WS (Won Stage) + Archive : "Leave Won"
	Later, if set P = 0% -> trigger "Reach Lost".
- In WS + Archive : Reach Won
	If proba set to 0 -> Should never happen
- Out of WS + Restore: Leave Won
- In WS + Restore: Reach Won
	If proba was 0 : Leave Lost

This commit adds two fields on crm.lead to ease domains to display/hide 
fields, buttons and ribbons:
- is_lost: computed on active=False AND proba = 0. Tracking enbaled to follow
	the state of the lead in the chatter. (when it was lost or restored)
- is_won: related to the stage_id.won_stage. Used to know if the lead is won.
	As won = in won stage.

Review the Restore vs Unarchive:
- Unarchive = set active=True
- Restore = Unarchive + force automated probability (and it's recomputation).
Restore button is displayed only on lost leads.

Probabilty is set to readonly mode when the lead is won or lost. This is to
avoid inconsistencies between the state of the lead and it's probability.

Why oh why ?
============

Before this change, the old won/lost rules where the following:
Lost: Proba = 0 OR archived
Won: won stage AND active

Which means that both rules could have coexisted. If the lead were in won stage,
active but probability = 0%. (that should not happens but was still allowed).
The objective with the new rules is to use the probability as a necessary
but not sufficient criteriat. Probability makes it impossible to be both
lost and won. But probability is not sufficient to classify a lead as won or
lost.
0% does not mean it's lost, it must also be archived.
100% does not mean it's won, it must also be in won stage.

That will avoid having automatically lost or won leads.
If Predictive lead scoring computes 0% or 100% chance of winning, the lead will
not be flagged yet as lost or won and the user will still need to take action to
correctly classify the lead.

Task-2654916

### CODE REVIEW

File: addons/crm/tests/test_crm_pls.py [hunk 1]
```diff
@@ -359,6 +395,7 @@ def test_predictive_lead_scoring(self):
 
         # Continue to test probability computation
         leads[3].probability = 40
+        leads[3].invalidate_recordset()
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_pls.py:398]: Should we test the edge cases as well?

Like for example writing 50 probability on a lead that is in "stage won" (the famous "not supposed to happen").

**Review by awa-odoo — COMMENTED** [review]
Hello @dbeguin 

Good news, I have finally read it!

...bad news, I have a lot of annoying questions :smile: 

Sorry if some of them are irrelevant.
I feel like we are trying to be "based on rules that are not guaranteed" in some cases.

Cheers!

File: addons/crm/models/crm_lead.py [hunk 2]
```diff
@@ -850,19 +862,35 @@ def _handle_won_lost(self, vals):
...
+            if vals.get('probability', 1) == 0 and not lead.active and not lead.stage_id.id in won_stage_ids:
+                leads_reach_lost += lead
+
+            # Leaving lost is a bit tricky as writing active = True and probability = x
+            # will not be done at same time in most cases. As lost need both conditions (active=False AND P=0), changing
+            # one of the two should trigger 'leave lost'. But this should not be triggered a second time if we modify
+            # the other condition. So the lead should leave lost only if both conditions are valid before writing and
+            # not after (if lead is lost before writing and not after).
+            # But this is only valid outside of the won_stage as the lead has already been removed from lost when
+            # arriving in won stage.
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/crm/models/crm_lead.py:886]: Hello @dbeguin 

Thought about this yesterday evening (yes, not sure why) and had another idea.
Would it be possible, with our brand new shiny computed fields, to go for something like the following:

```py
old_status_by_lead = {
    'is_won': lead.is_won,
    'is_lost': lead.is_lost
    for lead in self
}

super().write(self, vals)

new_status_by_lead = {
    'is_won': lead.is_won,
    'is_lost': lead.is_lost
    for lead in self
}

self._handle_won_lost(old_status_by_lead, new_status_by_lead)
```

Then in "_handle_won_lost", you would just have to check "was it lost before and not anymore -> leads_leave_lost", etc etc.

Do you see the idea?

Maybe it's stupid / non performant, but I thought it was worth the proposal.

Cheers!

File: addons/crm/models/crm_lead.py [hunk 3]
```diff
@@ -2282,16 +2347,14 @@ def _update_automated_probabilities(self):
...
-                    ('stage_id', '!=', False), ('create_date', '>=', pls_start_date),
-                '|',
-                    ('probability', '=', False),
-                    '&',
-                        ('probability', '<', 100), ('probability', '>', 0)
+            ('stage_id', '!=', False),
+            ('create_date', '>=', pls_start_date),
+            '|',
+                ('probability', '=', False),
+                ('is_won', '=', False),  # If active and P=0, we want to recompute?
```
Comment (Odoonan) [inline addons/crm/models/crm_lead.py:2357]: Here, we only recompute probas (when manually triggering the update) of Active leads, that are nor won nor lost.

However, those leads are updated in batch mode below and probabilities obtained with _pls_get_naive_bayes_probabilities(batch_mode=True)) called on them.

This is the only time batch_mode parameter is True. Then inside that method we find this:

```python
  # Get all leads values, no matter the team_id
  domain = []
  if batch_mode:
      domain = [
          '&',
              ('active', '=', True), ('id', 'in', self.ids),
              '|',
                  ('probability', '=', None),
                  '&',
                      ('probability', '<', 100), ('probability', '>', 0)
      ]
```

So two things: a bit of duplication (active = True), but maybe it is a security check?
And it seems like not recomputing 0 and 100 proba set manually is a choice. (it could happen in new heuristics: active with 0 or 100 without won stage) Do we want that?

**Review by Odoonan — COMMENTED** [review]
Hello @awa-odoo  Small technical review attached :)
Cheers!

File: addons/crm/models/crm_lead.py [hunk 4]
```diff
@@ -1900,15 +1968,15 @@ def _creation_subtype(self):
...
             return self.env.ref('crm.mt_lead_won')
         elif 'lost_reason_id' in init_values and self.lost_reason_id:
             return self.env.ref('crm.mt_lead_lost')
         elif 'stage_id' in init_values:
             return self.env.ref('crm.mt_lead_stage')
-        elif 'active' in init_values and self.active:
+        elif 'is_lost' in init_values and not self.is_lost:
             return self.env.ref('crm.mt_lead_restored')
-        elif 'active' in init_values and not self.active:
+        elif 'is_lost' in init_values and self.is_lost:
```
Comment (Odoonan) [inline addons/crm/models/crm_lead.py:1979]: have to double-check

File: addons/website_crm_partner_assign/controllers/main.py [hunk 5]
```diff
@@ -107,7 +107,7 @@ def portal_my_opportunities(self, page=1, date_begin=None, date_end=None, sortby
             'week': {'label': _('This Week Activities'),
                      'domain': [('activity_date_deadline', '>=', today), ('activity_date_deadline', '<=', this_week_end_date)]},
             'overdue': {'label': _('Overdue Activities'), 'domain': [('activity_date_deadline', '<', today)]},
-            'won': {'label': _('Won'), 'domain': [('stage_id.is_won', '=', True)]},
+            'won': {'label': _('Won'), 'domain': [('stage_id.is_won', '=', True), ('probability', '=', 100)]},
             'lost': {'label': _('Lost'), 'domain': [('active', '=', False), ('probability', '=', 0)]},
```
Comment (awa-odoo) [inline addons/website_crm_partner_assign/controllers/main.py:111]: Could be simplified "in_won" = True / "is_lost" = True ?

**Review by awa-odoo — COMMENTED** [review]
Posting the comments of our live review session.

Thanks & courache :muscle:

**Comment by Odoonan** [comment]
Hello @awa-odoo ,

I've updated the code to match the last discussion we had, and double checked the tests too. I think it's getting better.

I've also left in a separate commit the suggestion you made on using old_status new_status around the write call, and after the create call, for you to have a look. In that commit are also two domains (on lead duplicate detection and lead assignment) I've updated and could use some double check :p

Cheers!

File: addons/crm/models/crm_stage.py [hunk 6]
```diff
@@ -48,3 +48,21 @@ def default_get(self, fields):
...
         self.team_count = self.env['crm.team'].search_count([])
+
+    def write(self, vals):
+        """ Since leads that are in a won stage must have their
+        probability = 100%, this override ensures that setting a stage as won
+        will set all the lead in that stage to probability = 100%.
+        Inversely, if a won stage is not marked as won anymore, the lead
+        probability should be recomputed base on automated probability.
+        Note: If a user set a stage as won and change his mind right after,
+        the manuel probability will be lost in the process."""
```
Comment (awa-odoo) [inline addons/crm/models/crm_stage.py:59]: Few typos here:

base -> based
a user set -> a user sets
and change -> and changes
the manuel -> the manual

File: addons/crm/tests/test_crm_lead.py [hunk 7]
```diff
@@ -809,6 +809,75 @@ def test_phone_mobile_update(self):
...
+
+        # Won lead can be inactive
+        lead.write({'active': False})
+        self.assertTrue(lead.is_won)
+        with self.assertRaises(ValidationError, msg='A won lead cannot have a probability different than 100'):
+            lead.write({'probability': 50})
+
+        # Restore the lead in a non won stage. won_count = lost_count = 0.1 in frequency table. P = 50%
+        lead.write({'stage_id': stage_in_progress.id, 'active': True})
+        self.assertFalse(lead.probability == 100)
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_lead.py:857]: Could test that "is_won" is False here, right?

File: addons/crm/tests/test_crm_pls.py [hunk 8]
```diff
@@ -277,6 +277,8 @@ def test_predictive_lead_scoring(self):
 
         # Restore -> Should decrease lost
         leads[4].toggle_active()
+        self.assertEqual(leads[4].is_lost, False)
+        self.assertEqual(leads[4].is_won, False)
```
Comment (awa-odoo) [inline addons/crm/tests/test_crm_pls.py:281]: Could assertTrue / assertFalse for those (to apply here below as well).

File: addons/crm/views/crm_lead_views.xml [hunk 9]
```diff
@@ -77,8 +76,9 @@
                                         <field class="mb-0" name="automated_probability" force_save="1"/> %
                                     </small>
                                     <div id="probability" class="d-flex align-items-baseline">
+                                        <field name="is_won" invisible="1"/>
```
Comment (awa-odoo) [inline addons/crm/views/crm_lead_views.xml:79]: Could be put next to "is_lost" at the beginning of the sheet tag for better :socks:

File: addons/crm/views/crm_lead_views.xml [hunk 10]
```diff
@@ -567,10 +568,9 @@
                             <ul class="oe_kanban_colorpicker" data-field="color"/>
                         </t>
                         <t t-name="kanban-box">
-                            <t t-set="lost_ribbon" t-value="!record.active.raw_value and record.probability and record.probability.raw_value == 0"/>
+                            <t t-set="lost_ribbon" t-value="record.is_lost.raw_value"/>
```
Comment (awa-odoo) [inline addons/crm/views/crm_lead_views.xml:571]: Would say we don't really need a variable anymore, we could just make it conditioned based on "record.is_lost.raw_value", right?

File: addons/crm/views/crm_lead_views.xml [hunk 11]
```diff
@@ -643,7 +644,7 @@
                     <attribute name="recurring_revenue_sum_field">recurring_revenue_monthly_prorated</attribute>
                 </xpath>
                 <xpath expr="//t[@t-set='lost_ribbon']" position="after">
-                    <t t-set="won_ribbon" t-value="record.active.raw_value and record.probability and record.probability.raw_value == 100"/>
+                    <t t-set="won_ribbon" t-value="record.is_won.raw_value"/>
```
Comment (awa-odoo) [inline addons/crm/views/crm_lead_views.xml:647]: Same here, would say extra variable is unnecessary.

File: addons/crm/views/crm_lead_views.xml [hunk 12]
```diff
@@ -1010,8 +1010,8 @@
                     <filter string="Creation Date" name="creation_date" date="create_date"/>
                     <filter string="Closed Date" name="close_date" date="date_closed"/>
                     <separator/>
-                    <filter string="Won" name="filter_won" domain="['&amp;', ('active', '=', True), ('stage_id.is_won', '=', True)]"/>
-                    <filter string="Ongoing" name="filter_ongoing" domain="['&amp;', ('active', '=', True), ('stage_id.is_won', '=', False)]"/>
+                    <filter string="Won" name="filter_won" context="{'active_test': False}" domain="[('is_won', '=', True)]"/>
+                    <filter string="Ongoing" name="filter_ongoing" domain="['&amp;', ('is_lost', '=', False), ('is_won', '=', False)]"/>
                     <filter string="Lost" name="filter_lost" domain="['&amp;', ('active', '=', False), ('probability', '=', 0)]"/>
```
Comment (awa-odoo) [inline addons/crm/views/crm_lead_views.xml:1015]: Can we not make this based on "is_lost", "=", True instead now?

File: addons/crm/models/crm_lead.py [hunk 13]
```diff
@@ -541,7 +541,7 @@ def _search_is_lost(self, operator, value):
     @api.depends('stage_id.is_won', 'probability')
     def _compute_is_won(self):
         for lead in self:
-            lead.is_won = lead.probability == 100 and lead.stage_id.is_won
+            lead.is_won = lead.stage_id.is_won and lead.probability == 100
```
Comment (awa-odoo) [inline addons/crm/models/crm_lead.py:544]: Strange change inside this commit ^^

Will not matter if we squash commits, but I think keeping probability first is better for performance, as if it's not 100%, the code will not need to fetch the stage_id at all.

File: addons/crm/models/crm_lead.py [hunk 14]
```diff
@@ -816,20 +819,34 @@ def write(self, vals):
         elif stage_updated and not stage_is_won and not 'probability' in vals:
             vals['date_closed'] = False
 
-        if any(field in ['active', 'stage_id', 'probability'] for field in vals):
-            self._handle_won_lost(vals)
+        do_frequencies_update = any(field in ['active', 'stage_id', 'probability'] for field in vals)
```
Comment (awa-odoo) [inline addons/crm/models/crm_lead.py:822]: Maybe simply "update_frequencies" ?

File: addons/crm/models/crm_team.py [hunk 15]
```diff
@@ -382,8 +382,7 @@ def _allocate_leads(self, work_days=1):
         Heuristic of this method is the following:
           * find unassigned leads for each team, aka leads being
             * without team, without user -> not assigned;
-            * not in a won stage, and not having False/0 (lost) or 100 (won)
-              probability) -> live leads;
+            * not won nor lost -> live leads;
```
Comment (awa-odoo) [inline addons/crm/models/crm_team.py:385]: Note that the domain does not seem to match this, as it only excludes won?

I guess we rather ignore inactive leads, which also makes sense, but is not exactly the same.

**Review by awa-odoo — COMMENTED** [review]
Hello @Odoonan 

I left a few very minor additional comments while re-reading but it looks super clean now.

I much prefer the new logic inside your second commit, what do you think about it? We could probably squash both commits if we keep that one in the end.

We should be very close to merge now, may need to ask Thibault for a last check if he wants to.
I know the leads are a usual source of performance issue, but looking at the few domain changes, they seem relatively minor (and pretty much identical to what we had before), so hopefully it will be fine :crossed_fingers: 

Again, thank you for pushing this one to the end!

File: addons/crm/models/crm_lead.py (hunk 13 above)
Comment (Odoonan) [reply addons/crm/models/crm_lead.py]: Hello @awa-odoo 

I'm done, just posting a reply to this very comment. Actually, the query counters went up in two tests linked to _action_assign_leads. (1176 > 1167 and 6303 > 6066) When trying to switch stage and proba, it did not, hence the change.

I guessed this had to do with the probability being a computed field, even though stored. As we have queries in the pls flow, _compute_probabilities could lead to extra queries in comparison to reading the stage. I can investigate more if needed.

What do you think? I can at least already move it to Testing BE?

Cheers!

Comment (awa-odoo) [reply addons/crm/models/crm_lead.py]: Fine then, just seemed strange in that commit context, but indeed computed field can add queries so switching the check order actually matters.

Greened on my side, thanks again! 🟢

File: addons/crm/models/crm_lead.py [hunk 16]
```diff
@@ -860,40 +911,52 @@ def search_fetch(self, domain, field_names, offset=0, limit=None, order=None):
...
-        leads_reach_lost._pls_increment_frequencies(to_state='lost')
-        leads_leave_lost._pls_increment_frequencies(from_state='lost')
+        leads_reach_won_ids = []
+        leads_leave_won_ids = []
+        leads_reach_lost_ids = []
+        leads_leave_lost_ids = []
+
+        for lead_id, new_status in new_status_by_lead.items():
+            old_status = old_status_by_lead.get(lead_id, {'is_lost': False,
+                                                          'is_won': False})
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:939]: Would use "self" to be a real recordset based method, like
```suggestion
        for lead in self:
            new_status = new_status_by_lead.get(
                lead.id, {'is_lost': False, 'is_won': False}
            )
            old_status = old_status_by_lead.get(
                lead.id, {'is_lost': False, 'is_won': False}
            )
```
```

File: addons/crm/models/crm_lead.py [hunk 17]
```diff
@@ -860,40 +911,52 @@ def search_fetch(self, domain, field_names, offset=0, limit=None, order=None):
...
+
+            if new_status['is_lost'] and new_status['is_won']:
+                raise ValidationError(_("The lead %s cannot be won and lost at the same time.", Lead.browse(lead_id)))
+
+            if new_status['is_lost'] and not old_status['is_lost']:
+                leads_reach_lost_ids.append(lead_id)
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:945]: Would use recordsets, and do "leads_reach_lost += lead", if we use real records :)

File: addons/crm/tests/common.py [hunk 18]
```diff
@@ -158,14 +158,19 @@ def setUpClass(cls):
...
             'type': 'lead',
             'user_id': cls.user_sales_leads.id,
             'team_id': cls.sales_team_1.id,
         })
         cls.lead_team_1_lost.action_set_lost()
         (cls.lead_team_1_won + cls.lead_team_1_lost).flush_recordset()
 
+        # make lead 1 take team history into account for its automated proba.
+        # it should now be 50% as auto proba. (1 lost 1 won for team 1)
+        cls.lead_1._compute_probabilities()
```
Comment (tde-banana-odoo) [inline addons/crm/tests/common.py:171]: Shouldn't it be automatically be done if we flush ?

File: addons/crm/models/crm_stage.py [hunk 19]
```diff
@@ -48,3 +48,21 @@ def default_get(self, fields):
...
+        probability = 100%, this override ensures that setting a stage as won
+        will set all the leads in that stage to probability = 100%.
+        Inversely, if a won stage is not marked as won anymore, the lead
+        probability should be recomputed based on automated probability.
+        Note: If a user sets a stage as won and changes his mind right after,
+        the manual probability will be lost in the process."""
+        res = super().write(vals)
+        if 'is_won' in vals:
+            won_leads = self.env['crm.lead'].search([('stage_id', 'in', self.ids)])
+            if won_leads and vals.get('is_won'):
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_stage.py:63]: Can potentially be violent ... shouldn't we have a warning if somehow someone changes the flag ?

File: addons/crm/models/crm_lead.py [hunk 20]
```diff
@@ -860,40 +911,52 @@ def search_fetch(self, domain, field_names, offset=0, limit=None, order=None):
...
+                leads_reach_won_ids.append(lead_id)
+            elif not new_status['is_won'] and old_status['is_won']:
+                leads_leave_won_ids.append(lead_id)
+
+        Lead.browse(leads_reach_won_ids)._pls_increment_frequencies(to_state='won')
+        Lead.browse(leads_leave_won_ids)._pls_increment_frequencies(from_state='won')
+        Lead.browse(leads_reach_lost_ids)._pls_increment_frequencies(to_state='lost')
+        Lead.browse(leads_leave_lost_ids)._pls_increment_frequencies(from_state='lost')
+
+        return leads_reach_won_ids, leads_leave_won_ids, leads_reach_lost_ids, leads_leave_lost_ids
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py:959]: Is the return really necessary ? Does not seems used

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the PR and your work on it :) made a first reading, added few comments on coding style mainly. Will read again soon to be sure about the changes, but already looking good on my side.

Cheers !

**Review by Odoonan — COMMENTED** [review]
Hello @tde-banana-odoo 

Comments answered / code updated :). Also, regarding enterprise fields and possibly removing won_status:

- won_status in crm.lead 
This is stored. This is used in spreadsheet and we would need to update them + find a way to use is_lost and is_won instead. 

- won_status in crm.activity.report
This could probably be dealt with by adapting the _join, _select, ... methods 

In the spreadsheet, we have several pivots using won_status field as rowGroupBys. This means that the user can use pivots and specify the field with one of its values as the subset of records they consider to compute the measure. It is also the case in the value of some of the existing cells. (e.g.: 'Data' sheet)

Unfortunately, I am not sure we can do something similar using is_won and is_lost. (Or we would have to override the read_group_groupby for this field ? But even then, not sure how to go from "Pivot _id_, __count, won_status, 'pending'" to something similar using other fields, at least without virtually mapping values through the read_group_groupby, and I am not sure this would still work).

And we would also have to upgrade spreadsheets, as the field would not exist anymore. (I do not know if specific tools exists to do so)



Another simpler idea would be to bring won_status in crm module, hence removing is_lost and is_won (+ their compute and search methods), simplifying the shmilblick and using only this field everywhere.

I guess we did not use it so far in base crm module because it was explicitely only used in the crm_enterprise reporting and filters / spreadsheets? Do you know why we did not have a won_status historically? Probably just to avoid a stored field in the base crm module, I would guess.

What do you think? I could also give it a go in a separate commit and see how it turns out?

We could also keep won_status and just edit the compute method to match new heuristics. But that seems redundant.

Cheers & thanks for the follow-up!

**Comment by tde-banana-odoo** [comment]
> Hello @tde-banana-odoo
> 
> [...] Do you know why we did not have a won_status historically? Probably just to avoid a stored field in the base crm module, I would guess.

As it was considered as costly, we did not add it directly in CRM but only for modules really using it, indeed. I think we should try to have a single field for the status, to simplify and avoid data duplication (especially lead is a huge table). Note that it would be better to have is_won and is_lost fields, instead of a selection one (easier to store True / False I would say). Or an integer-based selection field (0 / 1 / 2), not sure if still supported by odoo ? Purpose would be to have a minimal impact on db :) .

> 
> What do you think? I could also give it a go in a separate commit and see how it turns out?
> 
> We could also keep won_status and just edit the compute method to match new heuristics. But that seems redundant.
> 
> Cheers & thanks for the follow-up!
